### DISCUSSION TITLE
Generating Secret for Custom EdgeX Device Service without using ADD_SECRETSTORE_TOKENS environment variable

### DISCUSSION THREAD

**Discussion by satyam1990** [edgexfoundry/community#150] [Q&A]
Hi Team,

I have a custom Device Service written using EdgeX C SDK, it starts or stops dynamically as and when required, hence it is not feasible for us to keep its name in ADD_SECRETSTORE_TOKENS environment variable.

Also this service makes use of REDIS message bus and we cannot also configure ADD_KNOWN_SECRETS: redisdb[MY_SERVICE_NAME], since service might not be running on a machine and hence we don't want these secrets to be present when my device service is not running.

Assume that the secretstore-setup is already running without containing name of my device service in env. variable ADD_SECRETSTORE_TOKENS. 

And now at a later point in time I want to start my custom device service then how should I get the secret tokens for my device service such that my device service can access Redis also.

**Comment by bnevis-i** [comment]
> I have a custom Device Service written using EdgeX C SDK, it starts or stops dynamically as and when required, hence it is not feasible for us to keep its name in ADD_SECRETSTORE_TOKENS environment variable. ...
> And now at a later point in time I want to start my custom device service then how should I get the secret tokens for my device service such that my device service can access Redis also.

There is a little-used feature in EdgeX designed exactly for this situation.  It is called [delayed-start services](https://docs.edgexfoundry.org/3.0/security/Ch-DelayedStartServices/).  Rather than pre-seeding your service with its secret store token. it uses workload attestation to prove the service's identity at runtime (e.g. by the service name of the container, or the container image hash, etc) to request a fresh secret store token on the fly.

** Unfortunately, the C SDK does not support the delayed-start feature (contributions accepted!) **  Though you would be able to write a small utility executable using the Go SDK, get the token, and then invoke your C service directly afterwards.

> Also this service makes use of REDIS message bus and we cannot also configure ADD_KNOWN_SECRETS: redisdb[MY_SERVICE_NAME], since service might not be running on a machine and hence we don't want these secrets to be present when my device service is not running.

This is required configuration and the purpose of the EdgeX secret store -- to protect these secrets at rest.  There is only one Redis user (this is a carryover from previous Redis versions that only supported a default user and a single password).  Every service has a copy of the redis secret, and that secret is inaccessible without a secret store token.  Making yet another copy of the redis secret for your service, in the secret store where all the other copies of the redis secret are stored, poses minimal additional risk.
